metasynthesis/abstract_genetic.py

```diff
@@ -0,0 +1,70 @@
+from typing import List, Callable, Tuple
+
+Genome = List
+Population = List[Genome]
+PopulateFunc = Callable[[], Population]
+FitnessFunc = Callable[[Genome], float]
+SelectionFunc = Callable[[Population, FitnessFunc], Tuple[Genome, Genome]]
+CrossoverFunc = Callable[[Genome, Genome], Tuple[Genome, Genome]]
+MutationFunc = Callable[[Genome], Genome]
+
+
+class GeneticAlgorithm:
+    """Abstract interface for a genetic algorithm to synthesise elements used in program synthesis"""
+
+    def __init__(self, fitness_limit: int, generation_limit: int,
+                 crossover_probability: float, mutation_probability: float):
+        self.fitness_limit = fitness_limit
+        self.generation_limit = generation_limit
+        self.crossover_probability = crossover_probability
+        self.mutation_probability = mutation_probability
+
+    def generate_genome(self, length: int) -> Genome:
+        """This method creates a new genome of the specified length"""
+
+        raise NotImplementedError()
+
+    def generate_population(self, size: int) -> Population:
+        """This method creates a population of new genomes"""
+
+        raise NotImplementedError()
+
+    def fitness(self, genome: Genome) -> float:
+        """This method calculates the fitness of the specified genome"""
+
+        raise NotImplementedError()
+
+    def single_point_crossover(self, a: Genome, b: Genome) -> Tuple[Genome, Genome]:
+        """This method combines first half of genome a with second half of b and vice versa with certain probability"""
+
+        raise NotImplementedError()
+
+    def n_point_crossover(self, a: Genome, b: Genome) -> Tuple[Genome, Genome]:
```

Comment:
I think having a more general crossover method would be sufficient in this case, that way people can implement them the way they like in their implementation. Using those abstract methods we could also already implement a run_evolution cycle that only needs those methods to run.

Reply:
Good point, I've merged the crossover methods into one which has a crossover function parameter